eclipse/src/saros/editor/EditorActivityDelayer.java

```diff
@@ -0,0 +1,89 @@
+package saros.editor;
+
+import java.util.Objects;
+import org.apache.log4j.Logger;
+import org.eclipse.swt.widgets.Display;
+import saros.activities.EditorActivity;
+import saros.activities.TextEditActivity;
+import saros.activities.TextSelectionActivity;
+import saros.activities.ViewportActivity;
+import saros.session.AbstractActivityProducer;
+import saros.session.ISarosSession;
+import saros.ui.util.SWTUtils;
+import saros.util.StackTrace;
+
+/**
+ * This class is responsible for delaying and discarding certain activities that are normally
+ * generated frequently.
+ */
+// TODO Viewport activities are also fire to frequently, worst case is selecting text with mouse
+// while scrolling up or down
+
+final class EditorActivityDelayer extends AbstractActivityProducer {
+
+  private static final int DELAY = 100; // ms
+
+  private static final Logger log = Logger.getLogger(EditorActivityDelayer.class);
+
+  private final Display display;
+
+  private TextSelectionActivity currentSelectionActivity = null;
+
+  EditorActivityDelayer() {
+    display = SWTUtils.getDisplay();
+    Objects.requireNonNull(display, "CTOR invoked from non SWT Thread");
+  }
+
+  void start(final ISarosSession session) {
+    session.addActivityProducer(this);
+  }
+
+  void stop(final ISarosSession session) {
+    session.removeActivityProducer(this);
+  }
+
+  void fireActivity(final ViewportActivity activity) {
+    checkSWTAccess();
+    flushCurrentSelectionActivity();
+    super.fireActivity(activity);
+  }
+
+  void fireActivity(final TextSelectionActivity activity) {
```

Comment:
Will the behavior of the other `fireActivity(...)` methods change in the future?

Otherwise, you could consider just having one `fireActivity(IActivity)` method and differentiate between the activity types internally. This would allow you to avoid the code duplication for the other methods.

Reply:
I do not add fireActivity(IActivity) for a special reason. Other developers should think about optimization. If I would add such a method there is no reason to think about it. And on the other hand the activities have not changed since almost 10 years. But thanks, I have to check if the fireActivity(IActivity) method of the AbstractActivityProducer can be accessed.